### PR TITLE
Document the requirement that commits have a public email address

### PR DESCRIPTION
**Pull request by AaronBallman**
See https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223 for details about why this is important to the community.

Note, we currently have soft enforcement for this requirement in the form of a bot which posts comments letting patch authors know their email is private, so we're already setting expectations in practice; this PR is documenting those expectations for clarity.

### CODE REVIEW

**Comment by AaronBallman** [comment]
Another reason why this is potentially becoming more important in the relatively near future is because the proposed governance process intends to use GitHub APIs to verify contributors for voter eligibility and as a way to send ballots/voting instructions to folks. That's not what's driving this PR, but it is what got me motivated to post it.

**Comment by joker-eph** [comment]
Are we gonna "upgrade" the bot message from a soft message to more of a requirement?
Considering the current dynamic since I raised this, I just ignore the bot and merge PRs (like most people it seems) at the moment.

**Comment by AaronBallman** [comment]
> Are we gonna "upgrade" the bot message from a soft message to more of a requirement? Considering the current dynamic since I raised this, I just ignore the bot and merge PRs (like most people it seems) at the moment.

To be honest, I don't know. I figured we'd get the policy down on paper first, then we can post another RFC to see if we want to strengthen the enforcement of the policy further. Personally, I'm comfortable with the soft action so long as the person landing changes on behalf of someone is willing to take responsibility for reverts if necessary. But the RFC thread was mixed on how firm we should be with enforcement, so I didn't want to change behavior with this PR.

**Comment by asl** [comment]
We can certainly upgrade the bot message and make it a required check. However, this would cover only maybe half of the issue. The bot checks for email addresses used in commits. However, the email address is rewritten during "Squash & Merge" and it seems we neither cannot control nor prevent this.

Looks like the only solution is to disable direct merges at all? And delegate to some bot?

**Comment by Bryce-MW** [comment]
I had a similar issue at work where we wanted a check on the final commit created by "Squash & Merge" and not just the earlier commits. We didn't end up going with it but the solution I came up with was using GitHub merge queues and having just that one check enabled

**Comment by AaronBallman** [comment]
Ping on the documentation changes.

File: llvm/docs/DeveloperPolicy.rst [hunk 1]
```diff
@@ -86,6 +86,11 @@ that notices of confidentiality or non-disclosure cannot be respected.
 Making and Submitting a Patch
 -----------------------------
 
+The LLVM community requires contributors to have a public email address
```
Comment (rnk) [inline llvm/docs/DeveloperPolicy.rst:89]: Should this be the first paragraph of the section? How about putting it after the non-disclosure notice paragraph, or immediately before?

File: llvm/docs/DeveloperPolicy.rst [hunk 2]
```diff
@@ -86,6 +86,11 @@ that notices of confidentiality or non-disclosure cannot be respected.
...
+associated with their GitHub commits, so please ensure that "Keep my email
+addresses private" is disabled in your
+`account settings <https://github.com/settings/emails>`_.
```
Comment (rnk) [inline llvm/docs/DeveloperPolicy.rst:92]: I think it's worth elaborating on the motivation. Here's some suggested text:

The LLVM project uses email to communicate to contributors outside of the GitHub platform about their past contributions. Primarily, our buildbot infrastructure uses emails to contact contributors about build and test failures.

**Comment by dwblaikie** [comment]
Just a +1 for the general direction. The CoC Committee has encountered this problem recently too - no ability to privately communicate with a community member/contributor if they don't have a public email address associated with their github account.

**Review by llvm-beanz — APPROVED** [review]
I think this policy change is needed and great.

**Comment by AaronBallman** [comment]
I've updated based on the feedback from @rnk

**Review by rnk — APPROVED** [review]
Thanks for making the edit!

**Pull request merged**